File: app/Sources/LocalCaptionKit/TranscriptWriter.swift
import Foundation

public struct TranscriptSaveResult: Equatable {
    public let txtURL: URL
    public let jsonURL: URL
}

/// Writes a finished transcript to disk: the human `.txt` plus a machine-readable `.json`
/// sidecar (product decision §22.4 = yes). Filenames come from the start time with
/// ` (2)`, ` (3)`… collision suffixes (SPEC.md §12.1).
public enum TranscriptWriter {
    /// JSON sidecar shape — session metadata + the full segment list.
    private struct Sidecar: Encodable {
        let session_name: String
        let started_at: String
        let ended_at: String
        let duration_seconds: Int
        let segments: [TranscriptSegment]
    }

    public static func save(transcript: Transcript,
                            folder: URL,
                            sessionName: String,
                            start: Date,
                            end: Date,
                            durationSeconds: Int,
                            showTimestamps: Bool) throws -> TranscriptSaveResult {
        let fm = FileManager.default
        try fm.createDirectory(at: folder, withIntermediateDirectories: true)

        let base = resolveBase(folder: folder, stamp: TimeFormat.fileStamp(start))
        let txtURL = folder.appendingPathComponent(base + ".txt")
        let jsonURL = folder.appendingPathComponent(base + ".json")

        let text = transcript.fileText(sessionName: sessionName, start: start, end: end,
                                       durationSeconds: durationSeconds, showTimestamps: showTimestamps)
        try Data(text.utf8).write(to: txtURL, options: .atomic)

        let enc = JSONEncoder()
        enc.outputFormatting = [.prettyPrinted, .sortedKeys]
        let sidecar = Sidecar(session_name: sessionName,
                              started_at: TimeFormat.iso(start),
                              ended_at: TimeFormat.iso(end),
                              duration_seconds: durationSeconds,
                              segments: transcript.segments)
        try enc.encode(sidecar).write(to: jsonURL, options: .atomic)

        return TranscriptSaveResult(txtURL: txtURL, jsonURL: jsonURL)
    }

    /// Pick a base filename that collides with neither an existing `.txt` nor `.json`.
    static func resolveBase(folder: URL, stamp: String) -> String {
        let fm = FileManager.default
        func free(_ base: String) -> Bool {
            !fm.fileExists(atPath: folder.appendingPathComponent(base + ".txt").path)
            && !fm.fileExists(atPath: folder.appendingPathComponent(base + ".json").path)
        }
        if free(stamp) { return stamp }
        var n = 2
        while !free("\(stamp) (\(n))") { n += 1 }
        return "\(stamp) (\(n))"
    }
}

File: windows/src/LocalCaption.Core/Transcripts/TranscriptWriter.cs
using System.Text.Json;
using System.Text.Json.Serialization;
using LocalCaption.Core.Data;

namespace LocalCaption.Core.Transcripts;

public readonly record struct TranscriptSaveResult(string TxtPath, string JsonPath);

/// <summary>
/// Writes a finished transcript to disk: the human <c>.txt</c> plus a machine-readable
/// <c>.json</c> sidecar. Filenames come from the start time, with <c> (2)</c>, <c> (3)</c>…
/// collision suffixes (SPEC.md §12.1).
/// </summary>
/// <remarks>
/// SPEC-WINDOWS.md §9.1 requires these files to be <b>byte-identical</b> to the macOS
/// output: UTF-8 without a BOM, <c>\n</c> endings, the same header, the same sidecar keys.
/// That is what lets §17 verify parity with a plain diff, so nothing here may use
/// <see cref="Environment.NewLine"/> or a default UTF-8 encoder.
/// </remarks>
public static class TranscriptWriter
{
    /// <summary>Sidecar shape — session metadata plus the full segment list, keys sorted.</summary>
    private sealed record Sidecar(
        [property: JsonPropertyName("duration_seconds")] int DurationSeconds,
        [property: JsonPropertyName("ended_at")] string EndedAt,
        [property: JsonPropertyName("segments")] IReadOnlyList<TranscriptSegment> Segments,
        [property: JsonPropertyName("session_name")] string SessionName,
        [property: JsonPropertyName("started_at")] string StartedAt);

    private static readonly JsonSerializerOptions SidecarOptions = new() { WriteIndented = true };

    public static TranscriptSaveResult Save(Transcript transcript, string folder, string sessionName,
                                            DateTimeOffset start, DateTimeOffset end,
                                            int durationSeconds, bool showTimestamps)
    {
        Directory.CreateDirectory(folder);

        var baseName = ResolveBase(folder, TimeFormat.FileStamp(start));
        var txtPath = Path.Combine(folder, baseName + ".txt");
        var jsonPath = Path.Combine(folder, baseName + ".json");

        var text = transcript.FileText(sessionName, start, end, durationSeconds, showTimestamps);
        Files.WriteAllTextAtomic(txtPath, Files.Lf(text));

        var sidecar = new Sidecar(durationSeconds, TimeFormat.Iso(end), transcript.Segments,
                                  sessionName, TimeFormat.Iso(start));
        Files.WriteAllTextAtomic(jsonPath, Files.Lf(JsonSerializer.Serialize(sidecar, SidecarOptions)));

        return new TranscriptSaveResult(txtPath, jsonPath);
    }

    /// <summary>Pick a base filename colliding with neither an existing .txt nor .json.</summary>
    internal static string ResolveBase(string folder, string stamp)
    {
        bool Free(string name) =>
            !File.Exists(Path.Combine(folder, name + ".txt")) &&
            !File.Exists(Path.Combine(folder, name + ".json"));

        if (Free(stamp)) return stamp;
        var n = 2;
        while (!Free($"{stamp} ({n})")) n++;
        return $"{stamp} ({n})";
    }
}
